### ISSUE TITLE
Gene positions are sometimes wrong

### ISSUE THREAD

**Issue by abretaud** [GMOD/Apollo#1770]
Another problem while parsing the gff exported from apollo 2.0.7:

Sometimes the start and/or end position of a gene doesn't match the position of its single mRNA.
For example:

```
scaffold928	.	gene	37421	69779	.	+	.	ID=8ef9db54-7db5-4f98-982c-5788ebcaba39;date_last_modified=2017-09-23;status=Approved;date_creation=2017-06-23
scaffold928	.	mRNA	40438	65172	.	+	.	Parent=8ef9db54-7db5-4f98-982c-5788ebcaba39;ID=e054eee3-435a-4f90-b0ef-3f3d90423cd8;date_last_modified=2017-06-23;date_creation=2017-06-23
```

Out of 2866 annotated genes, I get a wrong start position on 74 genes, and a wrong end position on 69 genes (some genes have both wrong positions, but some have only one position wrong)

**Comment by abretaud** [comment]
A magical query to get all genes that have problems with their position (269 for me):

```
select p.unique_name, min(pl.fmin), max(pl.fmax), min(pl.strand), min(cl.fmin), max(cl.fmax), min(cl.strand) from feature p, feature_relationship r, feature c, feature_location pl, feature_location cl where p.class = 'org.bbop.apollo.Gene' and p.id = r.parent_feature_id and r.child_feature_id = c.id and c.class = 'org.bbop.apollo.MRNA' and p.id = pl.feature_id and c.id = cl.feature_id group by p.unique_name having (min(pl.fmin) != min(cl.fmin) or max(pl.fmax) != max(cl.fmax) or min(pl.strand) != min(cl.strand)) order by p.unique_name;
```

To get only the genes with problems on the fmin (62 for me):

```
select p.unique_name, min(pl.fmin), max(pl.fmax), min(pl.strand), min(cl.fmin), max(cl.fmax), min(cl.strand) from feature p, feature_relationship r, feature c, feature_location pl, feature_location cl where p.class = 'org.bbop.apollo.Gene' and p.id = r.parent_feature_id and r.child_feature_id = c.id and c.class = 'org.bbop.apollo.MRNA' and p.id = pl.feature_id and c.id = cl.feature_id group by p.unique_name having (min(pl.fmin) != min(cl.fmin)) order by p.unique_name;
```

To get only the genes with problems on the fmax (69 for me):

```
select p.unique_name, min(pl.fmin), max(pl.fmax), min(pl.strand), min(cl.fmin), max(cl.fmax), min(cl.strand) from feature p, feature_relationship r, feature c, feature_location pl, feature_location cl where p.class = 'org.bbop.apollo.Gene' and p.id = r.parent_feature_id and r.child_feature_id = c.id and c.class = 'org.bbop.apollo.MRNA' and p.id = pl.feature_id and c.id = cl.feature_id group by p.unique_name having (max(pl.fmax) != max(cl.fmax)) order by p.unique_name;
```

To get only the genes with problems with their strand (148 for me):

```
select p.unique_name, min(pl.fmin), max(pl.fmax), min(pl.strand), min(cl.fmin), max(cl.fmax), min(cl.strand) from feature p, feature_relationship r, feature c, feature_location pl, feature_location cl where p.class = 'org.bbop.apollo.Gene' and p.id = r.parent_feature_id and r.child_feature_id = c.id and c.class = 'org.bbop.apollo.MRNA' and p.id = pl.feature_id and c.id = cl.feature_id group by p.unique_name having (min(pl.strand) != min(cl.strand)) order by p.unique_name;
```

**Comment by abretaud** [comment]
And an ugly update query to fix corrupted genes:

```
UPDATE feature_location pl
SET
    fmin = new_pos.fmin,
    fmax = new_pos.fmax,
    strand = new_pos.strand
FROM  (
    select
        p.id as id, 
        min(cl.fmin) as fmin, 
        max(cl.fmax) as fmax, 
        min(cl.strand) as strand
    from
        feature p, feature_relationship r, feature c, feature_location pl, feature_location cl
    where
            p.class = 'org.bbop.apollo.Gene'
        and p.id = r.parent_feature_id 
        and r.child_feature_id = c.id 
        and c.class = 'org.bbop.apollo.MRNA' 
        and p.id = pl.feature_id 
        and c.id = cl.feature_id 
    group by p.id
    having (
           min(pl.fmin) != min(cl.fmin) 
        or max(pl.fmax) != max(cl.fmax) 
        or min(pl.strand) != min(cl.strand))
   ) AS new_pos
where new_pos.id = pl.feature_id;
```

Use at your own risk :D

**Comment by deepakunni3** [comment]
Interesting observation. I'll see how I can reproduce this bug and patch a fix for it.

**Comment by nathandunn** [comment]
@abretaud  This will happen if a gene has multiple isoforms (the gene dimensions accompany all isoforms).   Do you have the full GFF3 for that gene?

**Comment by abretaud** [comment]
Cool, thanks for the quick fix! I'll update our instances and let you know if I still see some similar things occuring

I'm worried about people who host apollo instances, do you plan to automatically fix the corrupted genes when upgrading to next apollo release?

**Comment by nathandunn** [comment]
@abretaud  This gets tricky.   A lot of users typically run a validation script before export and this hadn't come up before.   I'm not sure if they hadn't looked, or various other method calls triggered the updateGene call.  

I think the best solution would be to take what you have and create two shell scripts.   One for detecting and one for "fixing".   In theory doing an "undo / redo" should fix it, as well.  

Any thoughts @deepakunni3 ?

**Comment by deepakunni3** [comment]
To fix existing annotations, we can expose an endpoint `annotationEditorService/updateGeneBoundaries` and have a script that supplies all genes to that endpoint. This operation will snap the gene boundaries to the boundaries of its longest isoform.

I'll create a PR for this soon.

**Comment by abretaud** [comment]
I think this issue should be reopened: I applied the patch on our instance as soon as you committed it, but a few genes have this problem even though they were created after applying the fix.

I think it's more frequent as I only have 3 genes on ~2000 new annotated genes, but still the problem is there.

For one gene, it was created as a big gene, then in the history there is a long list of delete exons/set exon boundaries/merge transcript (probably with an isoform, no significant change in gene length). I think the probem comes from a final "delete exon" that resulted in a much shorter gene.

For another one, it started as a small gene, then it got merged with another nearby transcript, then after some "set exon boundaries", there are a few "delete exon" that reduced it's size

The last gene is a single exon gene, with big changes in size due to "set exon boundaries", and then a final "flip strand" change. The wrong coordinate comes from an old start position before a "set exon boundaries" that moved the start position

So I guess there are some things to check on "delete exon" and "set exon boundaries" actions?
